### PR TITLE
DOC: TEST.rst: add example with `pytest.mark.parametrize`

### LINKED ISSUE #23283: DOC/TST: potential updates to NumPy Testing Guidelines?
### Issue with current documentation:

The [Tests on random data](https://numpy.org/doc/stable/reference/testing.html#tests-on-random-data) section of the [Testing Guidelines](https://numpy.org/devdocs/reference/testing.html) does not recommend the use of `np.random.Generator`. Also, [Creating many similar tests](https://numpy.org/devdocs/reference/testing.html#creating-many-similar-tests) does not mention the use of `pytest.parametrize`.

### Idea or request for content:

Are the recommendations of these sections still considered best practices?

If not, should the "Tests on random data" section be updated to recommend the use of `np.random.default_rng`? (Are Python `random` and the singleton `RandomState` still acceptable in new tests? Are common seeds like `0` OK?)

Should the "Creating many similar tests" section be updated to demonstrate the use of `parametrize`?
<details>

For example, replace:
```python3
class LinalgTestCase:
    def test_single(self):
        a = array([[1., 2.], [3., 4.]], dtype=single)
        b = array([2., 1.], dtype=single)
        self.do(a, b)

    def test_double(self):
        a = array([[1., 2.], [3., 4.]], dtype=double)
        b = array([2., 1.], dtype=double)
        self.do(a, b)
```
with
```python3
class LinalgTestCase:
    @pytest.mark.parametrize('dtype', [single, double])
    def test_dtypes(self):
        a = array([[1, 2], [3, 4]], dtype=dtype)  # or random data?
        b = array([2, 1], dtype=dtype)
        self.do(a, b)
```
?
(Also, what is `imply` further down in the example?)

</details>

I'd be happy to submit a PR, but I thought I'd ask first.

### PR DESCRIPTION
**Pull request by mdhaber**
The "Creating many similar tests" section of `TEST.rst` did not mention the use of `pytest.mark.parametrize`, which is a common and convenient way to run several similar tests. I also noticed the following about the example:

- All instructions of `test_single` and `test_double` were identical aside from the dtype. This did not seem consistent with the intent of the example.
- The example showed use of `np.matrix`, which is no longer recommended.
- The example used a function `imply`, which was not defined.

This commit removes the "Creating many similar tests" section and expands the parametric tests section with an example that uses `pytest.mark.parametrize`. It also mentions NumPy's `assert_equal`, `assert_allclose`, and `assert_less` functions and makes other minor adjustments.

Closes gh-23283

### CODE REVIEW

File: doc/TESTS.rst [hunk 1]
```diff
@@ -114,27 +114,41 @@ a test class::
...
-whether a certain assumption is valid. If the assertion fails, the test fails.
+Within these test methods, the ``assert`` statement or a specialized assertion
+function is used to test whether a certain assumption is valid. If the
+assertion fails, the test fails. Common assertion functions include:
+
+- :func:`numpy.testing.assert_equal` for testing exact elementwise equality
+  between a result array and a reference,
+- :func:`numpy.testing.assert_allclose` for testing near elementwise equality
+  between a result array and a reference (i.e. with specified relative and
+  absolute tolerances), and
```
Comment (seberg) [inline doc/TESTS.rst:125]: Maybe we should advertise the `_nulp` versions more?

File: doc/TESTS.rst [hunk 2]
```diff
@@ -114,27 +114,41 @@ a test class::
...
+- :func:`numpy.testing.assert_array_less` for testing (strict) elementwise
+  ordering between a result array and a reference.
+
+Note that these assertion functions only compare the numerical vales of the
```
Comment (seberg) [inline doc/TESTS.rst:129]:
```suggestion
Note that these assertion functions only compare the numerical values of the
```

Comment (seberg) [reply doc/TESTS.rst:129]: We now have `strict=True` on some (`assert_equal`?) to check a bit more (mainly the shape and probably the dtype).

File: doc/TESTS.rst [hunk 3]
```diff
@@ -114,27 +114,41 @@ a test class::
...
+which are more broadly used. These versions also accept a ``match``
+parameter, which should always be used to precisely target the intended
+warning or error.
 
 Note that ``test_`` functions or methods should not have a docstring, because
 that makes it hard to identify the test from the output of running the test
 suite with ``verbose=2`` (or similar verbosity setting).  Use plain comments
-(``#``) if necessary.
+(``#``) to describe the intent of the test and help the unfamiliar reader to
+interpret the code.
```
Comment (seberg) [inline doc/TESTS.rst:149]: Can't suggest over the additional line:  The last sentence can be removed, this is was a `nose` feature that it printed out the docstrings on failures, IIRC.  It doesn't matter anymore, I think.

**Review by seberg — APPROVED** [review]
Thanks, generally LGTM, its great to get rid of that weird class howto.

A few smaller comments, take or leave although it would be nice to see some of them addressed ;).

File: doc/TESTS.rst (hunk 2 above)
Comment (mdhaber) [reply doc/TESTS.rst:129]: Yes, `assert_array_equal`. But it's only that one, I think. How bout we add `strict` to the others and then I can mention it here?

File: doc/TESTS.rst (hunk 1 above)
Comment (mdhaber) [reply doc/TESTS.rst:125]: Maybe. But actually, what are NumPy 2.0's plans for the zoo of testing functions? Can the offerings be pared down to those that are recommended?

File: doc/TESTS.rst (hunk 3 above)
Comment (mdhaber) [reply doc/TESTS.rst:149]: Ok l'll remove that bit. Should it say to "Use a docstring (`"""`) or comments (`#`) to..." or shall I leave this sentence alone?

File: doc/TESTS.rst (hunk 1 above)
Comment (rgommers) [reply doc/TESTS.rst:125]: There is no such plan. The usage of older testing functions is too high to do that I'm afraid. The functions that were deprecated/removed typically had a usage count of a handful, in some cases up to ~30 or so, in SciPy, scikit-learn and Pandas. I just did a quick count for `assert_array_almost_equal` and it's 2838 in SciPy alone - so we can't touch it really.

This topic can be better addressed in the docs though. This PR is a great step in the right direction.

File: doc/TESTS.rst (hunk 3 above)
Comment (seberg) [reply doc/TESTS.rst:149]: Ah right, why not nudge to comments even if the large reason to avoid it is gone.

File: doc/TESTS.rst (hunk 1 above)
Comment (mdhaber) [reply doc/TESTS.rst:125]: OK. But how about listing `assert_array_equal` as "not recommended" following the plan suggested in https://github.com/numpy/numpy/issues/8457#issuecomment-1714230145?

File: doc/TESTS.rst (hunk 2 above)
Comment (ngoldbaum) [reply doc/TESTS.rst:129]: That PR got merged, care to update the text too?

**Comment by ngoldbaum** [comment]
This seems like a good improvement, just one comment but if you don't have time to update this is still a significant improvement over the status quo and should be merged.

File: doc/TESTS.rst (hunk 2 above)
Comment (mdhaber) [reply doc/TESTS.rst:129]: Sure, I'm happy to update this after adding `strict` to `assert_equal`.

Comment (ngoldbaum) [reply doc/TESTS.rst]: Ah missed that there was still another step, merging!

**Comment by ngoldbaum** [comment]
Thanks @mdhaber!

**Comment by mdhaber** [comment]
Squash merge (second commit is a one character touch-up without real commit message) would be great, and I'd be happy to come back and mention `strict` when I add it to the other functions, or ... oh, there it goes : )

**Pull request merged**